### PR TITLE
Multiscale training

### PR DESCRIPTION
**Pull request by RaphaelKreft**
- Added multi scale image loading to the oxparis dataset
- Different scales can be given in a list of scales
- scales can be chosen randomly or round robin

### CODE REVIEW

**Review by FelixMoeller3 — CHANGES_REQUESTED** [review]
We need to make sure that in a batch each image has the same size

File: gluefactory/datasets/oxford_paris_mini_1view_jpldd.py
Comment (hardik01shah) [inline gluefactory/datasets/oxford_paris_mini_1view_jpldd.py:49]: Gluefactory I think has a preprocessor class that we can directly use maybe? I we reshape to a fixed size we may end up squeezing the image which Remi suggested against doing. We should maintain the aspect ratio ideally i.e. the longer side should be the same after resizing like 512 (as we do in HPatches).

Here is the preprocessor class https://github.com/FelixMoeller3/glue-factory/blob/85180e039fd7f6434b0ee1363189cace3bd9a992/gluefactory/utils/image.py#L12

And here is the usage: https://github.com/FelixMoeller3/glue-factory/blob/85180e039fd7f6434b0ee1363189cace3bd9a992/gluefactory/datasets/hpatches.py#L94

I used this for some other dataloader for POLD2. Let me know if this makes sense and I can help integrate this.

Comment (RaphaelKreft) [reply gluefactory/datasets/oxford_paris_mini_1view_jpldd.py]: I implemented many changes that also reolve this conversation. I will add a comment below explaining in detail

**Comment by RaphaelKreft** [comment]
1. Added new scaling technique that preserves aspect ratio
2. Added square_pad option that must be activated in batched processing so that all images in same batch have same size
3. I use ImagePreprocessor for resizing and padding
4. I adapted joint_point_line_extractor to deal with padding in loss function
5. Multiscale learning is now implemented in oxparis and minidepth(partial)
6. Adapted merged_ds to allow multiscale learning

TODOs:
[] Add example configs for merged_ds training and multiscale learning
[] Test Merged DS with multiscale learning one minidepth gt is generated
[] If GT for minidepth is geenrated I need to finish minidepth and test it
[] If padding is used should I  add logic to remove keypoints that lie in padding area (altough this is unlikeli to happen)

**Comment by FelixMoeller3** [comment]
FTR: @RaphaelKreft is currently generating MiniDepth ground truth and will merge this after he has also tested it with MiniDepth

File: gluefactory/datasets/merge_datasets.py
Comment (FelixMoeller3) [inline gluefactory/datasets/merge_datasets.py:71]: What does num_selected_this_batch mean?

File: gluefactory/datasets/merge_datasets.py
Comment (FelixMoeller3) [inline gluefactory/datasets/merge_datasets.py:77]: I think this whole code block would be a lot simpler if the multi-scale configuration wasn't on the level of an individual dataset. Do you think it would be possible to implement this? If it's too complicated it's also fine by me for the purposes of the paper

File: gluefactory/datasets/oxford_paris_mini_1view_jpldd.py
Comment (FelixMoeller3) [inline gluefactory/datasets/oxford_paris_mini_1view_jpldd.py:392]: Couldn't we put this method and the few above in a parent class and reuse these methods that way? I feel like there is a lot of code duplication for each dataset

**Review by FelixMoeller3 — CHANGES_REQUESTED** [review]

File: gluefactory/datasets/merge_datasets.py
Comment (RaphaelKreft) [reply gluefactory/datasets/merge_datasets.py:71]: This variable represents the number of selected images that are already selected for the current batch.
It is needed to synchronize multi-scale-training between two datasets that are merged. Because every time we are switching to a new batch we potentially switch scales.

If needed I can also explain the full procedure. Lmk

File: gluefactory/datasets/merge_datasets.py
Comment (RaphaelKreft) [reply gluefactory/datasets/merge_datasets.py:77]: I think we need multiscale training also on single dataset level. At least for the ablation.

Or are you suggesting a change in implementation?

File: gluefactory/datasets/oxford_paris_mini_1view_jpldd.py
Comment (RaphaelKreft) [reply gluefactory/datasets/oxford_paris_mini_1view_jpldd.py:392]: I think that this could be an idea. but I would do it after paper submission as the time untiul submission can be used better. I will open an Issue for later once this is merged.

File: gluefactory/datasets/merge_datasets.py
Comment (FelixMoeller3) [reply gluefactory/datasets/merge_datasets.py:77]: Sure, I agree that we need it on a single dataset level. But maybe it is possible to override the single dataset configuration in case of having the merged dataset?

File: gluefactory/datasets/oxford_paris_mini_1view_jpldd.py
Comment (FelixMoeller3) [reply gluefactory/datasets/oxford_paris_mini_1view_jpldd.py:392]: Agreed

**Review by FelixMoeller3 — APPROVED** [review]

File: gluefactory/datasets/merge_datasets.py
Comment (RaphaelKreft) [reply gluefactory/datasets/merge_datasets.py:77]: You mean deactivating multiscale training for the datasets that are being merged and handle it in the merged_dataset class? 

This would require more logic on the merged_dataset class. In this case I tried to resuse logic as much as possible. But this could be a consideration for a refactoring when we design a comnmon multiscale training superclass :)

Will add this to the issue :)

**Pull request merged**
